### ISSUE TITLE
Lowering memory usage in AnnotateBamWithUMIs 

### ISSUE THREAD

**Issue by Poshi** [fulcrumgenomics/fgbio#705]
I saw that you are loading the entire FastQ file with the UMIs for processing. This makes sense when you expect a bam file with a different sorting, like an aligned and coordinate sorted bam.
But this could be streamlined with the alignment itself: you could get the output from the aligner (typically in sam or bam format), which is in the same order as the input, and attach the UMIs from the UMI FastQ, which, presumably, is the same sorting as the original FastQs. From the user point of view, it only needs to add a parameter like `--assume-same-sorting`.
That way, you could perform the UMI addition without having to pre-load the UMI FastQ and not using much memory.
You could even be able to process slices of the aligned file as long as the relative sorting is the same.
The simplified algorithm would be:
* read alignment
* search for first matching record in the UMI FastQ
* assign the found UMI to the alignment
* read next alignment and repeat

All of this taking into account the ability to read multiple UMI reads and setting them as the specs say (dash to separate UMI read sequences and spaces to separate UMI read qualities), but I think this is already in another open issue :-)

**Comment by nh13** [comment]
Agreed. This is certainly feasible and wouldn’t be too hard to implement. Happy to review a PR if you’re up for it.

**Comment by Poshi** [comment]
I'll try to create something useful, but I don't really think I would make it. Never coded in scala and I don't either have the environment for it. Although being a simple algorithm, not sure if I will be able to create it without a lot of work (learning a whole new language).
I could provide pseudocode, or code in another language (easily in Python, probably in C or even Java).
Let's see what can I do, but don't expect much :-/

**Comment by nh13** [comment]
It might be helpful to read through some other code if you're thinking of contributing, and Scala is just a more concise Java, so hopefully not too hard.  Unfortunately, we make such improvements on our own time, so prioritize paid client work (feel free to reach out if you'd like to engage with us).  Thank-you for your understanding!

**Comment by Poshi** [comment]
Just created a [PR](https://github.com/fulcrumgenomics/fgbio/pull/709) for this. I hope the code is not too horrible. I'm surprised I managed to do that in so few time given my null experience with scala.
